### PR TITLE
[mlir] Guard expensive string-based asserts behind EXPENSIVE_CHECKS

### PR DESCRIPTION
**Pull request by Mogball**
These string based asserts are important but they are extremely expensive. In compilation workloads with lots of interface checking, this can slow down as much as 3x.

### CODE REVIEW

File: mlir/include/mlir/IR/Attributes.h [hunk 1]
```diff
@@ -50,20 +50,17 @@ class Attribute {
...
+  [[deprecated("Use mlir::dyn_cast<U>() instead")]] U dyn_cast() const;
   template <typename U>
-  [[deprecated("Use mlir::dyn_cast_or_null<U>() instead")]]
-  U dyn_cast_or_null() const;
+  [[deprecated("Use mlir::dyn_cast_or_null<U>() instead")]] U
+  dyn_cast_or_null() const;
   template <typename U>
-  [[deprecated("Use mlir::cast<U>() instead")]]
-  U cast() const;
+  [[deprecated("Use mlir::cast<U>() instead")]] U cast() const;
```
Comment (joker-eph) [inline mlir/include/mlir/IR/Attributes.h:63]: Unrelated formatting change?

File: mlir/include/mlir/IR/Attributes.h [hunk 2]
```diff
@@ -303,7 +300,7 @@ class AttributeInterface
 protected:
   /// Returns the impl interface instance for the given type.
   static typename InterfaceBase::Concept *getInterfaceFor(Attribute attr) {
-#ifndef NDEBUG
+#ifdef EXPENSIVE_CHECKS
```
Comment (joker-eph) [inline mlir/include/mlir/IR/Attributes.h:303]: This is critical to check in asserts builds IMO. The promise is quite useless otherwise...

File: mlir/include/mlir/IR/Attributes.h (hunk 1 above)
Comment (Mogball) [reply mlir/include/mlir/IR/Attributes.h:63]: clang-format strikes again!

**Review by joker-eph — CHANGES_REQUESTED** [review]
I have strong concerns here about making the system unsafe and much harder to debug/develop.

File: mlir/include/mlir/IR/Attributes.h (hunk 2 above)
Comment (Mogball) [reply mlir/include/mlir/IR/Attributes.h:303]: I'm open to your opinion on this! On one hand, I agree. I opened this PR simply for discussion (and to evaluate the impact on mojo). Practically no one builds with EXPENSIVE_CHECKS by default. However, compared to asserts in, say, `isa` or `cast`, this one is realllyyyy expensive. What if we had a separate flag that was also on by default but toggleable? The problem is that because this is in a header file, it's subject to the NDEBUG flag of whoever includes the file, not the LLVM binaries themselves. It would be nice to specifically toggle these ones and other "heavy" asserts in MLIR off.

Comment (Mogball) [reply mlir/include/mlir/IR/Attributes.h:303]: Our compiler(s) are under development, so turning off asserts in downstream projects might not be as desirable. Having a slow compiler, however, is also affecting the developer experience.

Comment (joker-eph) [reply mlir/include/mlir/IR/Attributes.h:303]: > What if we had a separate flag that was also on by default but toggleable?

That seems fine to me.

> The problem is that because this is in a header file, it's subject to the NDEBUG flag of whoever includes the file, not the LLVM binaries themselves. 

Which is WAI I believe: they build in assert mode means they want to catch issues in their project, so this will catch issues with their uses of isa/...

Comment (stellaraccident) [reply mlir/include/mlir/IR/Attributes.h:303]: It's a balance. Subjectively, I think that asserts in milr are already over the line of being too expensive, but it is a fraught topic to bring up. I can't say whether this is one of the big culprits, but expensive checks in isa costing 3x of overall compile time is both troubling and would explain a lot based on my experience.

**Comment by Mogball** [comment]
I might come back to this one with a different set of flags.

**Comment by joker-eph** [comment]
Thanks Jeff! Having more flags to `mlir-config.h` seems like a good direction to me!
